blog/content/post/2023-06-07-twpm-spi-fix.md

```diff
@@ -0,0 +1,855 @@
+		}
+
+		state = STATE_PAYLOAD_TRANSFER;
+		break;
+
+	case STATE_PAYLOAD_TRANSFER:
+		HAL_DMA_Start_IT(hspi->hdmatx, (uint32_t)waitstate_insert, (uint32_t)&hspi->Instance->DR, sizeof waitstate_insert);
+		HAL_DMA_PollForTransfer(hspi->hdmarx, HAL_DMA_FULL_TRANSFER, HAL_MAX_DELAY);
+		HAL_DMA_Start_IT(hspi->hdmarx, (uint32_t)&hspi->Instance->DR, (uint32_t)header, sizeof header);
+
+		state = STATE_WAIT_HEADER;
+		break;
+	}
+}
+```
+
+I start the TX transfer first, then poll for RX DMA completion before
+re-programming the DMA channel. Now, the test succeeds.
+
+## Extending tests
+
+I have basic code that can read and write data over SPI, but I have tested only
+read of a zeroed register. Now, it is time to extend the tests so that we write
+random data of random lengths, then read the data back and check whether it is
+as expected. I already got shorter cables - 10 cm instead of 20 cm,
+and I have stable communication at 24 MHz:
+
+![](/img/stm32-spi-connections.jpg)
+
+I started with something simple
+
+```python
+tpm_write(0, bytes([1,2,3,4,5,6,7,8]))
+x = tpm_read(0, 8)
+assert x == [1,2,3,4,5,6,7,8]
+```
+
+and failed. The first transfer succeeded, but the second did not:
+
+![](/img/stm32-spi-failed-second-transfer.png)
+
+I hooked the debugger and saw that app was still polling for RX DMA completion.
+Looking again at the original code, I found that I incorrectly cleared
+`SPI_RXFIFO_THRESHOLD` bit - it should be clear for 16-bit frame length and set
+for 8-bit frame length.
+
+Changing
+
+```c
+CLEAR_BIT(hspi->Instance->CR2, SPI_RXFIFO_THRESHOLD);
+```
+
+to
+
+```c
+SET_BIT(hspi->Instance->CR2, SPI_RXFIFO_THRESHOLD);
+```
+
+solved the problem, however I got another one.
+
+![](/img/stm32-spi-readback-wrong-data.png)
+
+The wait state is properly inserted and terminated, but the payload is invalid.
+I split the test into two to pause the app between write and read from the
+register. Peeking at the `scratch_buffer` reveals that DMA went wrong, as the
+first three bytes were lost entirely.
+
+![](/img/stm32-scratch-state.png)
+
+Moreover, we are again stuck polling for DMA completion (DMA is still waiting
+for the remaining three bytes). The issue could be caused by too high delays
+between restarting of DMA transfers, so I lowered the SPI frequency to 100 KHz,
+but to my surprise, the result was exactly the same. I tested different data
+sizes, and the result was always the same (3 bytes lost). So, the
+`SPI_RXFIFO_THRESHOLD` fix only moved the problem a bit further. The outcome is
+still the same.
+
+## Summary
+
+That's all for this blog post. I got SPI working at 24 MHz when writing, but
+reading is broken. This is a significant improvement anyway. Further work will
+include fixing RX and implementing missing features, such as SPI bus aborts,
+SPI synchronization (using CS pin), and error recovery. Also, we plan to
+upstream SPI fixes to Zephyr.
```

Comment:
Where is the code from this task?
Do we already have some patches to Zephyr? Why not send them right away?

Reply:
I pushed the code to https://github.com/arturkow2/MySPITest and added link to the blogpost. We don't have patches right now, I did a small fix here which could be sent as PR after small modifications, but someone already merged that.